src/Formidable.js

```diff
@@ -538,17 +547,34 @@ class IncomingForm extends EventEmitter {
     return name;
   }
 
-  _setUpRename() {
-    const hasRename = typeof this.options.filename === 'function';
+  _joinDirectoryName(name) {
+    const newPath = path.join(this.uploadDir, name);
+
+    // prevent directory traversal attacks
+    if (!newPath.startsWith(this.uploadDir)) {
+      return path.join(this.uploadDir, this.options.defaultInvalidName);
+    }
 
-    if (this.options.keepExtensions === true && hasRename) {
-      this._rename = (part) => {
-        const resultFilepath = this.options.filename.call(this, part, this);
+    return newPath;
+  }
 
-        return this._uploadPath(part, resultFilepath);
+  _setUpRename() {
+    const hasRename = typeof this.options.filename === 'function';
+    if (hasRename) {
+      this._getNewName = (part) => {
+        let ext = '';
+        let name = this.options.defaultInvalidName;
+        if (part.filename) {
+          // can be null
+          ({ ext, name } = path.parse(part.filename));
+          if (!this.options.keepExtensions) {
+            ext = '';
+          }
+        }
+        return this.options.filename.call(this, name, ext, part, this);
       };
     } else {
-      this._rename = (part) => this._uploadPath(part);
+      this._getNewName = (part) => this._uploadPath(part);
```

Comment:
can move the `uploadPath` logic here direcly, cuz I don't like that `uploadPath` name anyway...